execution_engine/src/shared/scoped_guard.rs

```diff
@@ -0,0 +1,49 @@
+//! Implementation of scoped guards for various tasks.
+use std::{cell::RefCell, rc::Rc};
+
+/// When created it increases a counter, and decreases it when dropped.
+pub(crate) struct ScopedCountingGuard {
+    shared_value: Rc<RefCell<i32>>,
+}
+
+impl ScopedCountingGuard {
+    #[must_use]
+    pub(crate) fn new(shared_value: &Rc<RefCell<i32>>) -> Self {
+        shared_value.replace_with(|&mut value| value + 1);
```

Comment:
I addressed this. After thinking, I decided not to panic but have another error in case of overflow. I don't see any scenario where it would hit this error other than a programming error.

Reply:
But then I think we should panic rather than having the error variant, as it would be a programming error.